### PR TITLE
feat: add a HashTree type for the certificate state

### PR DESCRIPTION
**Pull request by hansl**
It understands labels and can lookup paths and return blobs of those
paths, if found.

### CODE REVIEW

**Review by ericswanson-dfinity — APPROVED** [review]
Thanks!

File: ic-agent/src/hash_tree.rs [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+    /// is not necessary.
+    fn lookup_label(&self, label: &Label) -> LookupLabelResult {
+        match self {
+            // If this node is a labeled node, check for the name. This assume a
+            HashTreeNode::Labeled(l, node) => match label.cmp(l) {
+                std::cmp::Ordering::Greater => LookupLabelResult::Continue,
+                std::cmp::Ordering::Equal => LookupLabelResult::Found(node.as_ref()),
+                // If this node has a smaller label than the one we're looking for, shortcut
+                // out of this search (sorted tree), we looked too far.
+                std::cmp::Ordering::Less => LookupLabelResult::Absent,
```
Comment (nomeata) [inline ic-agent/src/hash_tree.rs:310]: :angry: Github, just ate this long comment. Second try.

```suggestion
                std::cmp::Ordering::Less => LookupLabelResult::Error,
```
I don’t think you can claim `Absent ` here; the lookup codes doesn't yet do the necessary bookkeeping.

Consider this tree:
```
Fork( Fork(Labeled("a",…), Pruned(…)), Labeled("c", …) )
```
looking up label `b` in this tree should return `Unknown` (after all, the value _might_ be in the `Pruned` tree, can't tell).

I recommend to just return `Unknown` here for now, which is good enough to unblock this project; the algorithm can be refined in a later PR.

If you want to do it right now, here is the way to do it: While you traverse the tree in order, you pass a flag around `prev_item_was_pruned`, initially `false`. If you hit `Pruned`, you turn this to `true`, when you hit `Labeled` to `false`. When hitting a label that’s larger (or the end of the tree) than the searched for, return `Absent` if the flag was `false` and `Unknown` if it was `true`.

Comment (hansl) [reply ic-agent/src/hash_tree.rs]: I fixed this whole function after debugging with your examples. They helped a lot, thanks!

Comment (hansl) [reply ic-agent/src/hash_tree.rs]: For reference, it wasn't Error either. It really _is_ absent, but there is another state which is Unknown when we pass a Pruned on the way. It's just a bunch of edge cases that I fixed this morning essentially.

Comment (hansl) [reply ic-agent/src/hash_tree.rs]: >  While you traverse the tree in order, you pass a flag around prev_item_was_pruned

That's almost what I did. I have another LookupLabelResult value which is Unknown. Absent gets transformed into Unknown if it meets a pruned.

**Comment by hansl** [comment]
Merging this per @ericswanson-dfinity  request.

File: ic-agent/src/hash_tree.rs (hunk 1 above)
Comment (nomeata) [reply ic-agent/src/hash_tree.rs]: That works too, nice work!

**Pull request merged**